{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clustering 3k PBMCs following a Seurat Tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This started out (July 2017) with a demonstration that Scanpy would allow to reproduce most of Seurat's ([Satija *et al.*, 2015](https://doi.org/10.1038/nbt.3192)) clustering tutorial ([link](http://satijalab.org/seurat/pbmc3k_tutorial.html)), which we gratefully acknowledge. In the meanwhile, we have added and removed several pieces.\n",
    "\n",
    "The data consists in *3k PBMCs from a Healthy Donor* and is freely available from 10x Genomics ([here](http://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz) from this [webpage](https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.3.3 anndata==0.6.13 numpy==1.14.2 scipy==1.0.0 pandas==0.22.0 scikit-learn==0.19.1 statsmodels==0.9.0 python-igraph==0.7.1 louvain==0.6.1 \n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy.api as sc\n",
    "from scipy import sparse, io\n",
    "from collections import Counter\n",
    "import os.path\n",
    "from sklearn.metrics.cluster import adjusted_rand_score\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle\n",
    "import os\n",
    "import datetime\n",
    "import hyperopt\n",
    "from hyperopt import fmin, tpe, hp, Trials, STATUS_OK\n",
    "import time\n",
    "# import utils\n",
    "import seurat_utils\n",
    "import hyperopt_utils\n",
    "plt.ion()\n",
    "plt.show()\n",
    "sc.settings.set_figure_params(dpi=80)\n",
    "sc.settings.verbosity = 0  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "# sc.logging.print_versions()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "scanpy==1.3.3 anndata==0.6.13 numpy==1.14.2 scipy==1.0.0 pandas==0.22.0 scikit-learn==0.19.1 statsmodels==0.9.0 python-igraph==0.7.1 louvain==0.6.1 \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<module 'hyperopt_utils' from '/home/mada/ml/single-cell-sota/seurat/hyperopt_utils.py'>"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "import importlib\n",
    "importlib.reload(seurat_utils)\n",
    "importlib.reload(hyperopt_utils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Input datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "space = {\n",
    "    'load_inputDataset' : 'brainCIDR'\n",
    "    ,'bpreprocess_min_genes' : hp.choice('bpreprocess_min_genes', [True,False])\n",
    "    ,'preprocess_min_genes' : hp.quniform('preprocess_min_genes', 0, 500, 10)\n",
    "    \n",
    "    ,'bpreprocess_min_cells' : hp.choice('bpreprocess_min_cells', [True,False])\n",
    "    ,'preprocess_min_cells' : hp.quniform('preprocess_min_cells', 0, 20, 1)\n",
    "    \n",
    "    ,'bpreprocess_teta_total_features' : hp.choice('bpreprocess_teta_total_features', [True,False])\n",
    "    ,'preprocess_teta_total_features' : hp.quniform('preprocess_teta_total_features', 5000, 18000, 1000)\n",
    "\n",
    "    ,'preprocess_normalize_per_cell' : hp.choice('l_preprocess_normalize_per_cell', \n",
    "                                                 [True, False])\n",
    "\n",
    "    ,'bpreprocess_filter_min_mean' : hp.choice('bpreprocess_filter_min_mean', [True,False])\n",
    "    ,'preprocess_filter_min_mean' : hp.uniform('preprocess_filter_min_mean', 0, 0.5) # 0.0125\n",
    "    ,'preprocess_filter_max_mean' : hp.uniform('preprocess_filter_max_mean', 0, 6)\n",
    "    \n",
    "    ,'bpreprocess_filter_min_disp' : hp.choice('bpreprocess_filter_min_disp', [True,False])\n",
    "    ,'preprocess_filter_min_disp' : hp.uniform('preprocess_filter_min_disp', 0, 1)# 0.5\n",
    "    \n",
    "    ,'preprocess_regress_out' : hp.choice('l_preprocess_regress_out', ['none', 'total_counts'])\n",
    "    ,'preprocess_scale' : hp.choice('l_preprocess_scale', [-1, 5, 10, 15, 20]) #10\n",
    "    ,'cluster_n_neighbors' : hp.quniform('cluster_n_neighbors', 1, 20,1) #10\n",
    "    ,'cluster_n_pcs' : hp.quniform('cluster_n_pcs', 20, 50, 1)  # 40\n",
    "    ,'cluster_plot_pca' : False\n",
    "    ,'preprocess_plot_preprocessing' : False\n",
    "    ,'evaluate_plot_results': False\n",
    "    ,'preprocess_verbosity' : 0\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating new trials...\n"
     ]
    }
   ],
   "source": [
    "filename = 'deng_trials.pkl'   \n",
    "trials=hyperopt_utils.getTrials(filename ,restart = False )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating new trials...\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 5755 genes that are detected in less than 8.0 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 24\n",
      "    finished (0:00:03.11) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:01.81) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.05) --> found 12 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.48955034236836736\n",
      ">> Result: 0.48955034236836736\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 6722 genes that are detected in less than 11.0 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 28\n",
      "    finished (0:00:00.07) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.50) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.00) --> found 15 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.44734017754725597\n",
      ">> Result: 0.44734017754725597\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 6433 genes that are detected in less than 10.0 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 28\n",
      "    finished (0:00:00.13) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:01.07) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.13) --> found 7 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.5146106868880098\n",
      ">> Result: 0.5146106868880098\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 3625 genes that are detected in less than 4.0 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 47\n",
      "    finished (0:00:00.07) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.50) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.00) --> found 14 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.4634244068651716\n",
      ">> Result: 0.4634244068651716\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 4879 genes that are detected in less than 6.0 cells\n",
      "regressing out ['total_counts']\n",
      "    sparse input is densified and may lead to high memory use\n",
      "    finished (0:00:36.36)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 40\n",
      "    finished (0:00:00.16) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:01.02) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.03) --> found 8 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.6020025740567619\n",
      ">> Result: 0.6020025740567619\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 7192 genes that are detected in less than 13.0 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 29\n",
      "    finished (0:00:00.09) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.64) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.01) --> found 13 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.4513414317197487\n",
      ">> Result: 0.4513414317197487\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 32\n",
      "    finished (0:00:00.12) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.84) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.01) --> found 10 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.521423623433167\n",
      ">> Result: 0.521423623433167\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 29\n",
      "    finished (0:00:00.18) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:01.10) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.01) --> found 7 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.6138111061485612\n",
      ">> Result: 0.6138111061485612\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 2854 genes that are detected in less than 3.0 cells\n",
      "regressing out ['total_counts']\n",
      "    sparse input is densified and may lead to high memory use\n",
      "    finished (0:00:27.81)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 29\n",
      "    finished (0:00:00.19) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:01.06) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.01) --> found 7 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.5532842317122988\n",
      ">> Result: 0.5532842317122988\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "regressing out ['total_counts']\n",
      "    sparse input is densified and may lead to high memory use\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mada/anaconda3/lib/python3.6/site-packages/statsmodels/genmod/families/family.py:176: RuntimeWarning: invalid value encountered in true_divide\n",
      "  return np.sum(resid_dev * freq_weights * var_weights / scale)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 7388 genes that are detected in less than 14.0 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 31\n",
      "    finished (0:00:00.09) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.64) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.01) --> found 15 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.5142372442867872\n",
      ">> Result: 0.5142372442867872\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 38\n",
      "    finished (0:00:00.18) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:01.09) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.02) --> found 7 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.4301596373663228\n",
      ">> Result: 0.4301596373663228\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 35\n",
      "    finished (0:00:00.16) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:01.06) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.02) --> found 7 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.41499082446592533\n",
      ">> Result: 0.41499082446592533\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 7792 genes that are detected in less than 16.0 cells\n",
      "regressing out ['total_counts']\n",
      "    sparse input is densified and may lead to high memory use\n",
      "    finished (0:00:29.04)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 42\n",
      "    finished (0:00:00.07) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.40) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.00) --> found 18 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.40616818241082864\n",
      ">> Result: 0.40616818241082864\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 1807 genes that are detected in less than 2.0 cells\n",
      "regressing out ['total_counts']\n",
      "    sparse input is densified and may lead to high memory use\n",
      "    finished (0:00:01.97)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 25\n",
      "    finished (0:00:00.15) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.95) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.02) --> found 7 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.5508822626307036\n",
      ">> Result: 0.5508822626307036\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 7388 genes that are detected in less than 14.0 cells\n",
      "regressing out ['total_counts']\n",
      "    sparse input is densified and may lead to high memory use\n",
      "    finished (0:00:02.14)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 36\n",
      "    finished (0:00:00.10) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.79) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.01) --> found 10 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.5405909917505284\n",
      ">> Result: 0.5405909917505284\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mada/anaconda3/lib/python3.6/site-packages/scanpy/preprocessing/simple.py:1000: RuntimeWarning: invalid value encountered in true_divide\n",
      "  X /= scale\n",
      "/home/mada/anaconda3/lib/python3.6/site-packages/scanpy/preprocessing/simple.py:788: RuntimeWarning: invalid value encountered in greater\n",
      "  if max_value is not None: X[X > max_value] = max_value\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 50\n",
      "    finished (0:00:00.11) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.75) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.01) --> found 9 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.5436065512638625\n",
      ">> Result: 0.5436065512638625\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 6962 genes that are detected in less than 12.0 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 39\n",
      "    finished (0:00:00.13) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.87) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.01) --> found 10 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.6864264613571192\n",
      ">> Result: 0.6864264613571192\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 21\n",
      "    finished (0:00:00.10) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:00.70) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.01) --> found 12 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.47526314105991885\n",
      ">> Result: 0.47526314105991885\n",
      "... reading from cache file ./cache/input-brainCIDR-matrix.h5ad\n",
      "Loading dataset brainCIDR with 22085 genes and 420 cells\n",
      "filtered out 8187 genes that are detected in less than 18.0 cells\n",
      "regressing out ['total_counts']\n",
      "    sparse input is densified and may lead to high memory use\n",
      "    finished (0:00:13.19)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 27\n",
      "    finished (0:00:00.16) --> added to `.uns['neighbors']`\n",
      "    'distances', weighted adjacency matrix\n",
      "    'connectivities', weighted adjacency matrix\n",
      "computing UMAP\n",
      "    finished (0:00:01.08) --> added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "running Louvain clustering\n",
      "    using the \"louvain\" package of Traag (2017)\n",
      "    finished (0:00:00.02) --> found 7 clusters and added\n",
      "    'louvain', the cluster labels (adata.obs, categorical)\n",
      "Rand_index 0.671377329984993\n",
      ">> Result: 0.671377329984993\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def runSeurat(filename, space, max_evals = 2):\n",
    "    # Define function to optimise\n",
    "    def evaluateSeurat(args):\n",
    "        args['cluster_n_neighbors'] = int(args['cluster_n_neighbors'])\n",
    "        args['cluster_n_pcs'] = int(args['cluster_n_pcs'])\n",
    "\n",
    "        if args['bpreprocess_min_genes'] == False:\n",
    "            args['preprocess_min_genes'] = -1\n",
    "        if args['bpreprocess_min_cells'] == False:\n",
    "            args['preprocess_min_cells'] = -1\n",
    "        if args['bpreprocess_teta_total_features'] == False:\n",
    "            args['preprocess_teta_total_features'] = -1\n",
    "        if args['bpreprocess_filter_min_mean'] == False:\n",
    "            args['preprocess_filter_min_mean'] = -1\n",
    "        if args['bpreprocess_filter_min_disp'] == False:\n",
    "            args['preprocess_filter_min_disp'] = -1\n",
    "\n",
    "\n",
    "        if args['preprocess_teta_total_features'] is not None:\n",
    "            args['preprocess_teta_total_features'] = int(args['preprocess_teta_total_features'])\n",
    "        try:\n",
    "            resultDict = seurat_utils.run(args)\n",
    "        except:\n",
    "            return { 'status' : hyperopt.STATUS_FAIL}\n",
    "\n",
    "        print(f'>> Result: {resultDict[\"randIndex\"]}')\n",
    "        ret = {\n",
    "            'loss' : -resultDict['randIndex']\n",
    "            ,'status' : STATUS_OK\n",
    "            ,'eval_time' : time.time()        \n",
    "        }\n",
    "        return ret\n",
    "\n",
    "    trials = hyperopt_utils.getTrials(filename ,restart = False )\n",
    "    evals_per_epoch = 10\n",
    "    for e in range(len(trials), max_evals, evals_per_epoch):\n",
    "        best = fmin(evaluateSeurat\n",
    "                    ,space\n",
    "                    ,algo=tpe.suggest \n",
    "                    ,max_evals= e + evals_per_epoch\n",
    "                    ,trials=trials)\n",
    "\n",
    "        pickle.dump(trials, open(filename, 'wb'))\n",
    "    return trials\n",
    "\n",
    "trials = runSeurat(filename, space, max_evals = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>load_inputDataset</th>\n",
       "      <th>bpreprocess_min_genes</th>\n",
       "      <th>preprocess_min_genes</th>\n",
       "      <th>bpreprocess_min_cells</th>\n",
       "      <th>preprocess_min_cells</th>\n",
       "      <th>bpreprocess_teta_total_features</th>\n",
       "      <th>preprocess_teta_total_features</th>\n",
       "      <th>preprocess_normalize_per_cell</th>\n",
       "      <th>bpreprocess_filter_min_mean</th>\n",
       "      <th>preprocess_filter_min_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>preprocess_filter_min_disp</th>\n",
       "      <th>preprocess_regress_out</th>\n",
       "      <th>preprocess_scale</th>\n",
       "      <th>cluster_n_neighbors</th>\n",
       "      <th>cluster_n_pcs</th>\n",
       "      <th>cluster_plot_pca</th>\n",
       "      <th>preprocess_plot_preprocessing</th>\n",
       "      <th>evaluate_plot_results</th>\n",
       "      <th>preprocess_verbosity</th>\n",
       "      <th>result</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>True</td>\n",
       "      <td>190.0</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "      <td>False</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.484101</td>\n",
       "      <td>...</td>\n",
       "      <td>0.793795</td>\n",
       "      <td>none</td>\n",
       "      <td>5</td>\n",
       "      <td>7.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.489550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>350.0</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "      <td>False</td>\n",
       "      <td>15000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.482360</td>\n",
       "      <td>...</td>\n",
       "      <td>0.935533</td>\n",
       "      <td>none</td>\n",
       "      <td>20</td>\n",
       "      <td>4.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.447340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>100.0</td>\n",
       "      <td>True</td>\n",
       "      <td>10.0</td>\n",
       "      <td>True</td>\n",
       "      <td>16000.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.099535</td>\n",
       "      <td>...</td>\n",
       "      <td>0.481102</td>\n",
       "      <td>none</td>\n",
       "      <td>-1</td>\n",
       "      <td>14.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.514611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>250.0</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>True</td>\n",
       "      <td>18000.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.090160</td>\n",
       "      <td>...</td>\n",
       "      <td>0.476458</td>\n",
       "      <td>none</td>\n",
       "      <td>-1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.463424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>True</td>\n",
       "      <td>60.0</td>\n",
       "      <td>True</td>\n",
       "      <td>6.0</td>\n",
       "      <td>True</td>\n",
       "      <td>14000.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.193390</td>\n",
       "      <td>...</td>\n",
       "      <td>0.865988</td>\n",
       "      <td>total_counts</td>\n",
       "      <td>15</td>\n",
       "      <td>15.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.602003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>400.0</td>\n",
       "      <td>True</td>\n",
       "      <td>13.0</td>\n",
       "      <td>False</td>\n",
       "      <td>7000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.363020</td>\n",
       "      <td>...</td>\n",
       "      <td>0.760052</td>\n",
       "      <td>none</td>\n",
       "      <td>5</td>\n",
       "      <td>6.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.451341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>210.0</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>18000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.362870</td>\n",
       "      <td>...</td>\n",
       "      <td>0.111937</td>\n",
       "      <td>none</td>\n",
       "      <td>5</td>\n",
       "      <td>9.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.521424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>True</td>\n",
       "      <td>150.0</td>\n",
       "      <td>False</td>\n",
       "      <td>4.0</td>\n",
       "      <td>False</td>\n",
       "      <td>18000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.283363</td>\n",
       "      <td>...</td>\n",
       "      <td>0.171807</td>\n",
       "      <td>none</td>\n",
       "      <td>5</td>\n",
       "      <td>19.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.613811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>True</td>\n",
       "      <td>430.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>11000.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.337434</td>\n",
       "      <td>...</td>\n",
       "      <td>0.253619</td>\n",
       "      <td>total_counts</td>\n",
       "      <td>20</td>\n",
       "      <td>15.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.553284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>True</td>\n",
       "      <td>80.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>14000.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.149749</td>\n",
       "      <td>...</td>\n",
       "      <td>0.903608</td>\n",
       "      <td>total_counts</td>\n",
       "      <td>10</td>\n",
       "      <td>7.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>340.0</td>\n",
       "      <td>True</td>\n",
       "      <td>14.0</td>\n",
       "      <td>False</td>\n",
       "      <td>10000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.444797</td>\n",
       "      <td>...</td>\n",
       "      <td>0.948768</td>\n",
       "      <td>none</td>\n",
       "      <td>20</td>\n",
       "      <td>6.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.514237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>True</td>\n",
       "      <td>450.0</td>\n",
       "      <td>False</td>\n",
       "      <td>5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>16000.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.107025</td>\n",
       "      <td>...</td>\n",
       "      <td>0.837041</td>\n",
       "      <td>none</td>\n",
       "      <td>20</td>\n",
       "      <td>19.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.430160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>190.0</td>\n",
       "      <td>False</td>\n",
       "      <td>7.0</td>\n",
       "      <td>True</td>\n",
       "      <td>14000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.084154</td>\n",
       "      <td>...</td>\n",
       "      <td>0.558346</td>\n",
       "      <td>none</td>\n",
       "      <td>5</td>\n",
       "      <td>19.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.414991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>110.0</td>\n",
       "      <td>True</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.233427</td>\n",
       "      <td>...</td>\n",
       "      <td>0.460475</td>\n",
       "      <td>total_counts</td>\n",
       "      <td>-1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.406168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>110.0</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>9000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.384844</td>\n",
       "      <td>...</td>\n",
       "      <td>0.684016</td>\n",
       "      <td>total_counts</td>\n",
       "      <td>-1</td>\n",
       "      <td>15.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.550882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>True</td>\n",
       "      <td>270.0</td>\n",
       "      <td>True</td>\n",
       "      <td>14.0</td>\n",
       "      <td>False</td>\n",
       "      <td>8000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.491105</td>\n",
       "      <td>...</td>\n",
       "      <td>0.449455</td>\n",
       "      <td>total_counts</td>\n",
       "      <td>10</td>\n",
       "      <td>8.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.540591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>True</td>\n",
       "      <td>40.0</td>\n",
       "      <td>False</td>\n",
       "      <td>8.0</td>\n",
       "      <td>True</td>\n",
       "      <td>8000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.036709</td>\n",
       "      <td>...</td>\n",
       "      <td>0.042209</td>\n",
       "      <td>none</td>\n",
       "      <td>5</td>\n",
       "      <td>8.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.543607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>250.0</td>\n",
       "      <td>True</td>\n",
       "      <td>12.0</td>\n",
       "      <td>False</td>\n",
       "      <td>15000.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.171856</td>\n",
       "      <td>...</td>\n",
       "      <td>0.870026</td>\n",
       "      <td>none</td>\n",
       "      <td>-1</td>\n",
       "      <td>11.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.686426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>True</td>\n",
       "      <td>70.0</td>\n",
       "      <td>False</td>\n",
       "      <td>10.0</td>\n",
       "      <td>True</td>\n",
       "      <td>11000.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.139531</td>\n",
       "      <td>...</td>\n",
       "      <td>0.603891</td>\n",
       "      <td>none</td>\n",
       "      <td>15</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.475263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>brainCIDR</td>\n",
       "      <td>False</td>\n",
       "      <td>320.0</td>\n",
       "      <td>True</td>\n",
       "      <td>18.0</td>\n",
       "      <td>False</td>\n",
       "      <td>14000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.263938</td>\n",
       "      <td>...</td>\n",
       "      <td>0.185941</td>\n",
       "      <td>total_counts</td>\n",
       "      <td>15</td>\n",
       "      <td>15.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.671377</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   load_inputDataset bpreprocess_min_genes  preprocess_min_genes  \\\n",
       "0          brainCIDR                  True                 190.0   \n",
       "1          brainCIDR                 False                 350.0   \n",
       "2          brainCIDR                 False                 100.0   \n",
       "3          brainCIDR                 False                 250.0   \n",
       "4          brainCIDR                  True                  60.0   \n",
       "5          brainCIDR                 False                 400.0   \n",
       "6          brainCIDR                 False                 210.0   \n",
       "7          brainCIDR                  True                 150.0   \n",
       "8          brainCIDR                  True                 430.0   \n",
       "9          brainCIDR                  True                  80.0   \n",
       "10         brainCIDR                 False                 340.0   \n",
       "11         brainCIDR                  True                 450.0   \n",
       "12         brainCIDR                 False                 190.0   \n",
       "13         brainCIDR                 False                 110.0   \n",
       "14         brainCIDR                 False                 110.0   \n",
       "15         brainCIDR                  True                 270.0   \n",
       "16         brainCIDR                  True                  40.0   \n",
       "17         brainCIDR                 False                 250.0   \n",
       "18         brainCIDR                  True                  70.0   \n",
       "19         brainCIDR                 False                 320.0   \n",
       "\n",
       "   bpreprocess_min_cells  preprocess_min_cells  \\\n",
       "0                   True                   8.0   \n",
       "1                   True                  11.0   \n",
       "2                   True                  10.0   \n",
       "3                   True                   4.0   \n",
       "4                   True                   6.0   \n",
       "5                   True                  13.0   \n",
       "6                  False                   2.0   \n",
       "7                  False                   4.0   \n",
       "8                   True                   3.0   \n",
       "9                  False                   1.0   \n",
       "10                  True                  14.0   \n",
       "11                 False                   5.0   \n",
       "12                 False                   7.0   \n",
       "13                  True                  16.0   \n",
       "14                  True                   2.0   \n",
       "15                  True                  14.0   \n",
       "16                 False                   8.0   \n",
       "17                  True                  12.0   \n",
       "18                 False                  10.0   \n",
       "19                  True                  18.0   \n",
       "\n",
       "   bpreprocess_teta_total_features  preprocess_teta_total_features  \\\n",
       "0                            False                          5000.0   \n",
       "1                            False                         15000.0   \n",
       "2                             True                         16000.0   \n",
       "3                             True                         18000.0   \n",
       "4                             True                         14000.0   \n",
       "5                            False                          7000.0   \n",
       "6                             True                         18000.0   \n",
       "7                            False                         18000.0   \n",
       "8                             True                         11000.0   \n",
       "9                            False                         14000.0   \n",
       "10                           False                         10000.0   \n",
       "11                            True                         16000.0   \n",
       "12                            True                         14000.0   \n",
       "13                           False                          5000.0   \n",
       "14                            True                          9000.0   \n",
       "15                           False                          8000.0   \n",
       "16                            True                          8000.0   \n",
       "17                           False                         15000.0   \n",
       "18                            True                         11000.0   \n",
       "19                           False                         14000.0   \n",
       "\n",
       "   preprocess_normalize_per_cell bpreprocess_filter_min_mean  \\\n",
       "0                           True                        True   \n",
       "1                           True                       False   \n",
       "2                          False                       False   \n",
       "3                          False                        True   \n",
       "4                          False                       False   \n",
       "5                           True                        True   \n",
       "6                           True                        True   \n",
       "7                           True                        True   \n",
       "8                          False                        True   \n",
       "9                          False                       False   \n",
       "10                          True                       False   \n",
       "11                         False                        True   \n",
       "12                          True                        True   \n",
       "13                          True                       False   \n",
       "14                          True                        True   \n",
       "15                          True                        True   \n",
       "16                          True                       False   \n",
       "17                         False                       False   \n",
       "18                         False                        True   \n",
       "19                          True                        True   \n",
       "\n",
       "    preprocess_filter_min_mean    ...     preprocess_filter_min_disp  \\\n",
       "0                     0.484101    ...                       0.793795   \n",
       "1                     0.482360    ...                       0.935533   \n",
       "2                     0.099535    ...                       0.481102   \n",
       "3                     0.090160    ...                       0.476458   \n",
       "4                     0.193390    ...                       0.865988   \n",
       "5                     0.363020    ...                       0.760052   \n",
       "6                     0.362870    ...                       0.111937   \n",
       "7                     0.283363    ...                       0.171807   \n",
       "8                     0.337434    ...                       0.253619   \n",
       "9                     0.149749    ...                       0.903608   \n",
       "10                    0.444797    ...                       0.948768   \n",
       "11                    0.107025    ...                       0.837041   \n",
       "12                    0.084154    ...                       0.558346   \n",
       "13                    0.233427    ...                       0.460475   \n",
       "14                    0.384844    ...                       0.684016   \n",
       "15                    0.491105    ...                       0.449455   \n",
       "16                    0.036709    ...                       0.042209   \n",
       "17                    0.171856    ...                       0.870026   \n",
       "18                    0.139531    ...                       0.603891   \n",
       "19                    0.263938    ...                       0.185941   \n",
       "\n",
       "   preprocess_regress_out  preprocess_scale cluster_n_neighbors cluster_n_pcs  \\\n",
       "0                    none                 5                 7.0          24.0   \n",
       "1                    none                20                 4.0          28.0   \n",
       "2                    none                -1                14.0          28.0   \n",
       "3                    none                -1                 4.0          47.0   \n",
       "4            total_counts                15                15.0          40.0   \n",
       "5                    none                 5                 6.0          29.0   \n",
       "6                    none                 5                 9.0          32.0   \n",
       "7                    none                 5                19.0          29.0   \n",
       "8            total_counts                20                15.0          29.0   \n",
       "9            total_counts                10                 7.0          42.0   \n",
       "10                   none                20                 6.0          31.0   \n",
       "11                   none                20                19.0          38.0   \n",
       "12                   none                 5                19.0          35.0   \n",
       "13           total_counts                -1                 3.0          42.0   \n",
       "14           total_counts                -1                15.0          25.0   \n",
       "15           total_counts                10                 8.0          36.0   \n",
       "16                   none                 5                 8.0          50.0   \n",
       "17                   none                -1                11.0          39.0   \n",
       "18                   none                15                 7.0          21.0   \n",
       "19           total_counts                15                15.0          27.0   \n",
       "\n",
       "    cluster_plot_pca  preprocess_plot_preprocessing evaluate_plot_results  \\\n",
       "0              False                          False                 False   \n",
       "1              False                          False                 False   \n",
       "2              False                          False                 False   \n",
       "3              False                          False                 False   \n",
       "4              False                          False                 False   \n",
       "5              False                          False                 False   \n",
       "6              False                          False                 False   \n",
       "7              False                          False                 False   \n",
       "8              False                          False                 False   \n",
       "9              False                          False                 False   \n",
       "10             False                          False                 False   \n",
       "11             False                          False                 False   \n",
       "12             False                          False                 False   \n",
       "13             False                          False                 False   \n",
       "14             False                          False                 False   \n",
       "15             False                          False                 False   \n",
       "16             False                          False                 False   \n",
       "17             False                          False                 False   \n",
       "18             False                          False                 False   \n",
       "19             False                          False                 False   \n",
       "\n",
       "   preprocess_verbosity    result  \n",
       "0                     0 -0.489550  \n",
       "1                     0 -0.447340  \n",
       "2                     0 -0.514611  \n",
       "3                     0 -0.463424  \n",
       "4                     0 -0.602003  \n",
       "5                     0 -0.451341  \n",
       "6                     0 -0.521424  \n",
       "7                     0 -0.613811  \n",
       "8                     0 -0.553284  \n",
       "9                     0       NaN  \n",
       "10                    0 -0.514237  \n",
       "11                    0 -0.430160  \n",
       "12                    0 -0.414991  \n",
       "13                    0 -0.406168  \n",
       "14                    0 -0.550882  \n",
       "15                    0 -0.540591  \n",
       "16                    0 -0.543607  \n",
       "17                    0 -0.686426  \n",
       "18                    0 -0.475263  \n",
       "19                    0 -0.671377  \n",
       "\n",
       "[20 rows x 22 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = getResultsAsDf(trials, space)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['load_inputDataset'] = inputDatasets[0]\n",
    "params['preprocess_min_genes'] = 200\n",
    "params['preprocess_min_cells'] = 3\n",
    "# Define on the dataset\n",
    "params['preprocess_teta_total_features'] = None\n",
    "params['preprocess_normalize_per_cell'] = True\n",
    "params['preprocess_filter_min_mean'] = 0.0125\n",
    "params['preprocess_filter_max_mean']=3\n",
    "params['preprocess_filter_min_disp']=0.5\n",
    "params['preprocess_regress_out']=None\n",
    "params['preprocess_scale']=10\n",
    "\n",
    "params['cluster_n_neighbors'] = 10\n",
    "params['cluster_n_pcs'] = 40\n",
    "params['cluster_plot_pca'] = False\n",
    "\n",
    "params['preprocess_plot_preprocessing'] = False\n",
    "params['evaluate_plot_results'] = False\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# evaluate(params)\n",
    "print(evaluate(params))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {'cluster_n_neighbors': 8,\n",
    " 'cluster_n_pcs': 37,\n",
    " 'cluster_plot_pca': False,\n",
    " 'evaluate_plot_results': False,\n",
    " 'load_inputDataset': 'deng',\n",
    " 'preprocess_filter_max_mean': 0.01,\n",
    " 'preprocess_filter_min_disp': 0.2843934859859411,\n",
    " 'preprocess_filter_min_mean': 0.1,\n",
    " 'preprocess_min_cells': None,\n",
    " 'preprocess_min_genes': None,\n",
    " 'preprocess_normalize_per_cell': True,\n",
    " 'preprocess_plot_preprocessing': False,\n",
    " 'preprocess_regress_out': None,\n",
    " 'preprocess_scale': 20,\n",
    " 'preprocess_teta_total_features': None,\n",
    " 'preprocess_verbosity': 0}\n",
    "resultDict = seurat_utils.run(params)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## No total counts filter/ regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['load_inputDataset'] = inputDatasets[0]\n",
    "params['preprocess_min_genes'] = 200\n",
    "params['preprocess_min_cells'] = 3\n",
    "# Define on the dataset\n",
    "params['preprocess_teta_total_features'] = None\n",
    "params['preprocess_normalize_per_cell'] = True\n",
    "params['preprocess_filter_min_mean'] = 0.0125\n",
    "params['preprocess_filter_max_mean']=3\n",
    "params['preprocess_filter_min_disp']=0.5\n",
    "params['preprocess_regress_out']=None\n",
    "params['preprocess_scale']=10\n",
    "\n",
    "params['cluster_n_neighbors'] = 10\n",
    "params['cluster_n_pcs'] = 40\n",
    "params['cluster_plot_pca'] = False\n",
    "\n",
    "params['evaluate_plot_results'] = True\n",
    "resultDict = seurat_utils.run(params)\n",
    "results.append(resultDict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## No gene filtering based on dispertion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['load_inputDataset'] = inputDatasets[0]\n",
    "params['preprocess_min_genes'] = 200\n",
    "params['preprocess_min_cells'] = 3\n",
    "# Define on the dataset\n",
    "params['preprocess_teta_total_features'] = 14000\n",
    "params['preprocess_normalize_per_cell'] = True\n",
    "params['preprocess_filter_min_mean'] = None\n",
    "params['preprocess_filter_max_mean']=None\n",
    "params['preprocess_filter_min_disp']=None\n",
    "params['preprocess_regress_out']=None\n",
    "params['preprocess_scale']=10\n",
    "\n",
    "params['cluster_n_neighbors'] = 10\n",
    "params['cluster_n_pcs'] = 40\n",
    "params['cluster_plot_pca'] = False\n",
    "\n",
    "params['evaluate_plot_results'] = True\n",
    "resultDict = seurat_utils.run(params)\n",
    "results.append(resultDict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## No scale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['load_inputDataset'] = inputDatasets[0]\n",
    "params['preprocess_min_genes'] = 200\n",
    "params['preprocess_min_cells'] = 3\n",
    "# Define on the dataset\n",
    "params['preprocess_teta_total_features'] = 14000\n",
    "params['preprocess_normalize_per_cell'] = True\n",
    "params['preprocess_filter_min_mean'] = 0.0125\n",
    "params['preprocess_filter_max_mean']=3\n",
    "params['preprocess_filter_min_disp']=0.5\n",
    "params['preprocess_regress_out']='total_counts'\n",
    "params['preprocess_scale']=None\n",
    "\n",
    "params['cluster_n_neighbors'] = 10\n",
    "params['cluster_n_pcs'] = 40\n",
    "params['cluster_plot_pca'] = False\n",
    "\n",
    "params['evaluate_plot_results'] = True\n",
    "resultDict = seurat_utils.run(params)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results.append(resultDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## No min genes/cells filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['load_inputDataset'] = inputDatasets[0]\n",
    "params['preprocess_min_genes'] = None\n",
    "params['preprocess_min_cells'] = None\n",
    "# Define on the dataset\n",
    "params['preprocess_teta_total_features'] = None\n",
    "params['preprocess_normalize_per_cell'] = True\n",
    "params['preprocess_filter_min_mean'] =None\n",
    "params['preprocess_filter_max_mean']=None\n",
    "params['preprocess_filter_min_disp']=None\n",
    "params['preprocess_regress_out']=None\n",
    "params['preprocess_scale']=None\n",
    "\n",
    "params['cluster_n_neighbors'] = 10\n",
    "params['cluster_n_pcs'] = 40\n",
    "params['cluster_plot_pca'] = False\n",
    "\n",
    "params['evaluate_plot_results'] = True\n",
    "resultDict = seurat_utils.run(params)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results.append(resultDict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## No preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['load_inputDataset'] = inputDatasets[0]\n",
    "params['preprocess_min_genes'] = None\n",
    "params['preprocess_min_cells'] = None\n",
    "# Define on the dataset\n",
    "params['preprocess_teta_total_features'] = None\n",
    "params['preprocess_normalize_per_cell'] = False\n",
    "params['preprocess_filter_min_mean'] =None\n",
    "params['preprocess_filter_max_mean']=None\n",
    "params['preprocess_filter_min_disp']=None\n",
    "params['preprocess_regress_out']=None\n",
    "params['preprocess_scale']=None\n",
    "\n",
    "params['cluster_n_neighbors'] = 10\n",
    "params['cluster_n_pcs'] = 40\n",
    "params['cluster_plot_pca'] = False\n",
    "\n",
    "params['evaluate_plot_results'] = True\n",
    "resultDict = seurat_utils.run(params)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results.append(resultDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalization only, less clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['load_inputDataset'] = inputDatasets[0]\n",
    "params['preprocess_min_genes'] = None\n",
    "params['preprocess_min_cells'] = None\n",
    "# Define on the dataset\n",
    "params['preprocess_teta_total_features'] = None\n",
    "params['preprocess_normalize_per_cell'] = True\n",
    "params['preprocess_filter_min_mean'] =None\n",
    "params['preprocess_filter_max_mean']=None\n",
    "params['preprocess_filter_min_disp']=None\n",
    "params['preprocess_regress_out']=None\n",
    "params['preprocess_scale']=None\n",
    "\n",
    "params['cluster_n_neighbors'] = 8\n",
    "params['cluster_n_pcs'] = 30\n",
    "params['cluster_plot_pca'] = False\n",
    "\n",
    "params['evaluate_plot_results'] = True\n",
    "resultDict = seurat_utils.run(params)\n",
    "results.append(resultDict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalization only, more clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['load_inputDataset'] = inputDatasets[0]\n",
    "params['preprocess_min_genes'] = None\n",
    "params['preprocess_min_cells'] = None\n",
    "# Define on the dataset\n",
    "params['preprocess_teta_total_features'] = None\n",
    "params['preprocess_normalize_per_cell'] = True\n",
    "params['preprocess_filter_min_mean'] =None\n",
    "params['preprocess_filter_max_mean']=None\n",
    "params['preprocess_filter_min_disp']=None\n",
    "params['preprocess_regress_out']=None\n",
    "params['preprocess_scale']=None\n",
    "\n",
    "params['cluster_n_neighbors'] = 15\n",
    "params['cluster_n_pcs'] = 50\n",
    "params['cluster_plot_pca'] = False\n",
    "\n",
    "params['evaluate_plot_results'] = True\n",
    "resultDict = seurat_utils.run(params)\n",
    "results.append(resultDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter less cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['load_inputDataset'] = inputDatasets[0]\n",
    "params['preprocess_min_genes'] = 200\n",
    "params['preprocess_min_cells'] = 2\n",
    "# Define on the dataset\n",
    "params['preprocess_teta_total_features'] = 14000\n",
    "params['preprocess_normalize_per_cell'] = True\n",
    "params['preprocess_filter_min_mean'] = 0.01\n",
    "params['preprocess_filter_max_mean']=4\n",
    "params['preprocess_filter_min_disp']=0.5\n",
    "params['preprocess_regress_out']='total_counts'\n",
    "params['preprocess_scale']=15\n",
    "\n",
    "params['cluster_n_neighbors'] = 10\n",
    "params['cluster_n_pcs'] = 40\n",
    "params['cluster_plot_pca'] = False\n",
    "\n",
    "params['evaluate_plot_results'] = True\n",
    "resultDict = seurat_utils.run(params)\n",
    "results.append(resultDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['load_inputDataset'] = inputDatasets[0]\n",
    "params['preprocess_min_genes'] = 200\n",
    "params['preprocess_min_cells'] = 3\n",
    "# Define on the dataset\n",
    "params['preprocess_teta_total_features'] = 14000\n",
    "params['preprocess_normalize_per_cell'] = True\n",
    "params['preprocess_filter_min_mean'] = 0.0125\n",
    "params['preprocess_filter_max_mean']=3\n",
    "params['preprocess_filter_min_disp']=0.5\n",
    "params['preprocess_regress_out']='total_counts'\n",
    "params['preprocess_scale']=10\n",
    "\n",
    "params['cluster_n_neighbors'] = 10\n",
    "params['cluster_n_pcs'] = 40\n",
    "params['cluster_plot_pca'] = False\n",
    "\n",
    "params['evaluate_plot_results'] = True\n",
    "resultDict = seurat_utils.run(params)\n",
    "results.append(resultDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Render all combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame( results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "135.88235473632812px",
    "width": "252.35293579101562px"
   },
   "number_sections": false,
   "sideBar": false,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
